crates/typst/src/eval/mod.rs

```diff
@@ -1076,23 +1077,71 @@ impl Eval for ast::Binary<'_> {
     }
 }
 
-/// Apply a basic binary operation.
+impl<'a> Eval for ast::ChainedComp<'a> {
+    type Output = Value;
+
+    #[tracing::instrument(name = "ChainedComp::eval", skip_all)]
+    fn eval(self, vm: &mut Vm) -> SourceResult<Self::Output> {
+        let mut side = |lhs: &dyn Fn() -> ast::Expr<'a>,
+                        op,
+                        rhs: &dyn Fn() -> ast::Expr<'a>| match op {
+            ast::BinOp::Eq => apply_binary_op(lhs, op, rhs, self.span(), vm, ops::eq),
+            ast::BinOp::Neq => apply_binary_op(lhs, op, rhs, self.span(), vm, ops::neq),
+            ast::BinOp::Lt => apply_binary_op(lhs, op, rhs, self.span(), vm, ops::lt),
+            ast::BinOp::Leq => apply_binary_op(lhs, op, rhs, self.span(), vm, ops::leq),
+            ast::BinOp::Gt => apply_binary_op(lhs, op, rhs, self.span(), vm, ops::gt),
+            ast::BinOp::Geq => apply_binary_op(lhs, op, rhs, self.span(), vm, ops::geq),
+            _ => bail!(
+                self.span(),
+                "{} is not a valid comparison operator",
+                self.lhs_op().as_str()
+            ),
+        };
+
+        let lhs = side(&|| self.lhs(), self.lhs_op(), &|| self.mid())?;
+        let rhs = side(&|| self.mid(), self.rhs_op(), &|| self.rhs())?;
+        match (&lhs, &rhs) {
+            (Value::Bool(a), Value::Bool(b)) => Ok(Value::Bool(*a && *b)),
+
+            // This should not occur since the parser should prevent it.
+            _ => bail!(self.span(), "cannot compare {} and {}", lhs.ty(), rhs.ty()),
+        }
+    }
+}
+
+/// Apply a basic binary operation from a binary expression.
 fn apply_binary_expr(
     binary: ast::Binary,
     vm: &mut Vm,
     op: fn(Value, Value) -> StrResult<Value>,
 ) -> SourceResult<Value> {
-    let lhs = binary.lhs().eval(vm)?;
+    apply_binary_op(|| binary.lhs(), binary.op(), || binary.rhs(), binary.span(), vm, op)
+}
+
+/// Apply a basic binary operation.
+fn apply_binary_op<'a, L, R>(
```

Comment:
why all the closures here? why not just `&ast::Expr`?

Reply:
That's what I changed it too. I was under the (wrong) assumption that the reason why the `Binary` was passed to the closure was somehow for performance reason and I wanted to keep the getting of `lhs` or `rhs` as late as possible, but it doesn't make much sense in hindsight. That's what I get for opening a PR at like midnight 😅